src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java

```diff
@@ -761,23 +890,30 @@ protected SCMRevision retrieve(@NonNull final String revision, @NonNull final Ta
             context.wantTags(true);
             context.withoutRefSpecs();
         }
+        if (fullHashMatch != null) {
+            //since this would have been skipped if this was a head or a tag we can just return whatever
+            return new GitRefSCMRevision(new GitRefSCMHead(fullHashMatch, fullHashMatches.iterator().next()), fullHashMatch);
+        }
         if (shortHashMatch != null) {
             // woot this seems unambiguous
             for (String name: shortNameMatches) {
                 if (name.startsWith(Constants.R_HEADS)) {
                     listener.getLogger().printf("Selected match: %s revision %s%n", name, shortHashMatch);
                     // WIN it's also a branch
-                    return new SCMRevisionImpl(new SCMHead(StringUtils.removeStart(name, Constants.R_HEADS)),
+                    return new GitBranchSCMRevision(new GitBranchSCMHead(StringUtils.removeStart(name, Constants.R_HEADS)),
                             shortHashMatch);
+                } else if (name.startsWith(Constants.R_TAGS)) {
+                    tagName = StringUtils.removeStart(name, Constants.R_TAGS);
+                    context.wantBranches(false);
+                    context.wantTags(true);
+                    context.withoutRefSpecs();
                 }
```

Comment:
If it is a ref, we save effort by returning the `GitRefSCMRevision` itself, so we should set a boolean here to say "found a ref match" and then return the `GitRefSCMRevision` at the end of the loop if the ref match was found. the current code will prefer a tag to a `GitRefSCMRevision` despite a tag needing date resolution (i.e. more work)

Reply:
Yes that's the way I designed it after how I interpreted it from our earlier discussions, that it's better if we could return a tag or a branch than a ref.
I guess I misunderstood?

Reply:
a branch is the best
a ref should be preferred to a tag (because we do not need to resolve the time)